### ISSUE TITLE
Bundler.require when Gemfile.lock isn't present

### ISSUE THREAD

**Issue by quoideneuf** [bundler/bundler#4015]
I'm not sure if this behavior is intentional or not. (If not, I guess this is a an objection rather than a bug report).

If I have a gem in a non-default group like "test" and run

````
bundle install --without test
````

I get a gemset with just my defaults. If my app uses Bundler.require I would expect it to run fine without the test gems (i.e., if I'm bundling the app for distribution). However, this isn't the case if I don't include Gemfile.lock.

I wrote a simple script to illustrate this:

https://github.com/quoideneuf/ImBunderCurious

Thanks

**Comment by segiddins** [comment]
That's intentionally, you should be running `bundle install` before calling `Bundler.require`

**Comment by quoideneuf** [comment]
Right, but my script does that...

**Comment by segiddins** [comment]
And then you're removing the lockfile... Don't do that.

**Comment by quoideneuf** [comment]
So, does that mean that Bundler.require requires Gemfile.lock to be present? If that is the case, the overall behavior still seems counter-intuitive. Here I've added some lines to the script:

https://github.com/quoideneuf/ImBunderCurious/blob/master/script.sh#L23-L30

Also, if Gemfile.lock IS required for Bundler.require, why not have a message similar to the one you get when Gemfile is missing?
````
Could not locate Gemfile or .bundle/ directory (Bundler::GemfileNotFound)
````

**Comment by philnash** [comment]
I'm intrigued by this error. Could you post the error output from your example app? (I don't have rvm availlable right now.)

**Comment by quoideneuf** [comment]
@philnash Sure, here it is - thanks for taking a look:

https://github.com/quoideneuf/ImBunderCurious/blob/master/out.txt

**Comment by segiddins** [comment]
That is all the expected bundler behavior...

**Comment by quoideneuf** [comment]
If that is the case, I think it's rather un user-friendly. If presence of Gemfile.lock is a hard requirement for Bundler.require, it should fail the same way it does when Gemfile is absent. No?

**Comment by segiddins** [comment]
It actually isn't the lock file that's needed, but rather that all of the gems that running `bundle lock` would resolve to have to be installed

**Comment by quoideneuf** [comment]
Assuming that the Gemfile includes default and test groups, Bundle.require:

succeeds if all default and test gems are present, but Gemfile.lock is absent
succeeds if only default gems are present, and Gemfile.lock is present
fails if only default gems are present, but Gemfile.lock is absent

This just seems inherently confusing and non intuitive, design-wise, but I feel like I'm being pedantic and am going to drop it.

**Comment by indirect** [comment]
If we were going to be completely consistent, we would have to remove any ability to ever run `Bundler.require` without a Gemfile.lock.

In order to run `Bundler.require`, Bundler needs two things: 1) every gem that is not in a `without` group must be installed and 2) knowledge that every gem version in the entire bundle (including gems in `without` groups) do not have any conflicts. The second can only be provided by either a Gemfile.lock or by examining the gemspecs of every gem, including those that are in without groups.

If all default and test gems are present, Bundler is able to examine the gemspecs of all the gems. Both 1 and 2 are met. If only the default gems and also the lock is present, Bundler can use the lock to be sure that there are no version conflicts, and so is able to use the installed default gems. If neither the test gems or the lock is present, Bundler literally cannot asses whether your Gemfile is logically possible or not.

Because of these constraints, the Bundler documentation clearly states that `Gemfile.lock` should be checked into source control and included in all deployment or packaging situations.

In conclusion: this was not an explicit design decision, this is a case where you are deliberately creating a situation where it is impossible for Bundler to do its job (which is to guarantee that all the gems in your Bundle have no version conflicts). What are to trying to accomplish by deleting the lock, and why isn't it possible for you to run `bundle install` before using your Bundle?

**Comment by quoideneuf** [comment]
@indirect Thank you for explaining this, it helps me greatly to understand what's going on. I am still puzzled by `Bundler.require`'s requirement 2. Why does it need to assess the logical possibility of my Gemfile at runtime? The horse has already left the barn, so to speak. And, though the documentation indeed advises users to keep Gemfile.lock around, it also says this:

>If you have a lot of dependencies, you might want to say "require all of the gems in my Gemfile". To do this, put the following code immediately following require 'bundler/setup': 
>`Bundler.require(:default)`
>For our example Gemfile, this line is exactly equivalent to:
>`require 'rails'`
>`require 'rack-cache'`
>`require 'nokogiri'`

Just to explain the angle I'm coming from and arrived in this seemingly contrived situation: I work on a JRuby product that is distributed as as cluster of Java webapps and an embedded Jetty webserver. Each webapp is in its own `war` file with its own Gemfile, but all share a single set of packaged gems. For the past 3 years, it has been distributed without any Gemfile.lock files (I can't say exactly why) but our customers are also downloading all our test and development dependencies as part of the distribution. I wanted to change that and ended up here. Incidentally, there appears to be a separate issue - possibly Warbler related - with simply adding the lock files, so I need to work out what that is or just back out of `Bundler.require` entirely. Thanks again for the help.

**Comment by indirect** [comment]
@quoideneuf this is a conversation that we have every few weeks with people :/ the reason that Bundler needs either a lock or every gem (including those in `--without` groups) is because the entire point of Bundler is to guarantee that every version of every gem in the bundle is compatible with one another.

If you have a lockfile, Bundler can simply assume that the lockfile contains all of the versions that will work together, and so it is happy to install only some of those gems.

If you do not have a lockfile, Bundler must read the dependencies listed for every single gem, including those in `--without` groups. Without every single gemspec, Bundler cannot guarantee that a later installation that does include the `test` gems would be able to succeed. That is why you must either run `bundle install` or have the `test` gems installed on the machine.

Distributing the package without a Gemfile.lock just means that Bundler is creating a lock on the fly the first time that each Gemfile is loaded—it would even make your applications boot faster if you included it, because then Bundler wouldn't have to read every gemspec and do a dependency graph resolution.

**Comment by quoideneuf** [comment]
@indirect If you are having to have this conversation every few weeks, then I assume that there is a significant group of people who would be happier if they could type `Bundler.require :resolve => false`. What would it cost you to give that to them (us)? Is it safe to assume that this would be trivial for you to implement? I understand that guaranteeing gem compatibility is the central feature of Bundler, and what makes it interesting to work on, but nothing you've said has helped me understand why I, as a downstream user, am forced to 'benefit' from this capability at runtime if I use `Bundler.require`. And I can't believe that in the end you would be more beholden to a principle that to the vagaries of real-world usage. Not everyone is deploying their software by doing `git update && bundle install && rails s`.

**Comment by indirect** [comment]
@quoideneuf I don't think that's actually true—the other people that I have had this conversation with have always, as far as I can remember, said that they understand that Bundler needs to make sure that all the gems will work together without throwing exceptions, and we move on.

First, if it were possible to run `Bundler.require :resolve => false`, you might as well just be using RubyGems without Bundler, because that could throw a gem activation error anytime. Second, it's not possible to require a gem without knowing exactly what version of that gem to require—that means Bundler.require literally cannot run without the resolver running.

As I said earlier, the entire core of Bundler is written to guarantee that a set of gem versions are compatible. `Bundler.require` is just a nice side-effect that happens to be possible after the core Bundler code has run. What you're asking is for us to throw away the core of Bundler and only leave the side effect. That would require re-implementing the entire Bundler.require feature from scratch without depending on the resolver, and it's not going to happen.

I suggest that you either continue to ship the test and development gems in your production packages, or you start shipping the Gemfile.lock in your production packages. Bundler isn't going to be able to help you without one or the other.
